validator/rpc/handlers_keymanager.go

```
@@ -19,11 +21,290 @@ import (
+	if err != nil {
+		http2.HandleError(w, errors.Wrap(err, "Could not import keystores").Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// If any of the keys imported had a slashing protection history before, we
+	// stop marking them as deleted from our validator database.
+	http2.WriteJson(w, &ImportKeystoresResponse{Data: statuses})
+}
+
+// DeleteKeystores allows for deleting specified public keys from Prysm.
+func (s *Server) DeleteKeystores(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.DeleteKeystores")
+	defer span.End()
+
+	if s.validatorService == nil {
+		http2.HandleError(w, "Validator service not ready.", http.StatusServiceUnavailable)
+		return
+	}
+	if !s.walletInitialized {
+		http2.HandleError(w, "Prysm Wallet not initialized. Please create a new wallet.", http.StatusServiceUnavailable)
+		return
+	}
+	km, err := s.validatorService.Keymanager()
+	if err != nil {
+		http2.HandleError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	var req DeleteKeystoresRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http2.HandleError(w, "Could not decode request body: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+	if len(req.Pubkeys) == 0 {
+		http2.WriteJson(w, &DeleteKeystoresResponse{Data: make([]*keymanager.KeyStatus, 0)})
+		return
+	}
+	deleter, ok := km.(keymanager.Deleter)
+	if !ok {
+		sts := make([]*keymanager.KeyStatus, len(req.Pubkeys))
+		for i := 0; i < len(req.Pubkeys); i++ {
+			sts[i] = &keymanager.KeyStatus{
+				Status:  keymanager.StatusError,
+				Message: fmt.Sprintf("Keymanager kind %T cannot delete local keys", km),
+			}
+		}
+		http2.WriteJson(w, &DeleteKeystoresResponse{Data: sts})
+		return
+	}
+	bytePubKeys := make([][]byte, len(req.Pubkeys))
+	for i, pubkey := range req.Pubkeys {
+		key, ok := shared.ValidateHex(w, "Pubkey", pubkey, fieldparams.BLSPubkeyLength)
+		if !ok {
+			return
```

Missing `http2.HandleError`

---

it does it inside the validate Hex , maybe we should change that to be more aware? and return an error instead ?

---

Ah, I didn't notice the `w` param. Might be indeed better to return the error, but it should be a different PR.